File: packages/google_mlkit_language_id/ios/Classes/GoogleMlKitLanguageIdPlugin.m
#import "GoogleMlKitLanguageIdPlugin.h"
#import <MLKitLanguageID/MLKitLanguageID.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#define channelName @"google_mlkit_language_identifier"
#define startLanguageIdentifier @"nlp#startLanguageIdentifier"
#define closeLanguageIdentifier @"nlp#closeLanguageIdentifier"

@implementation GoogleMlKitLanguageIdPlugin {
    NSMutableDictionary *instances;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitLanguageIdPlugin* instance = [[GoogleMlKitLanguageIdPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startLanguageIdentifier]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:closeLanguageIdentifier]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (MLKLanguageIdentification*)initialize:(FlutterMethodCall *)call {
    NSNumber *confidence = call.arguments[@"confidence"];
    MLKLanguageIdentificationOptions *options = [[MLKLanguageIdentificationOptions alloc] initWithConfidenceThreshold:confidence.floatValue];
    return [MLKLanguageIdentification languageIdentificationWithOptions:options];
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *uid = call.arguments[@"id"];
    MLKLanguageIdentification *languageId = [instances objectForKey:uid];
    if (languageId == NULL) {
        languageId = [self initialize:call];
        instances[uid] = languageId;
    }
    
    BOOL possibleLanguages = [call.arguments[@"possibleLanguages"] boolValue];
    NSString *text = call.arguments[@"text"];
    if(possibleLanguages) {
        [self identifyPossibleLanguagesInText:text languageId:languageId result:result];
    } else {
        [self identifyLanguageInText:text languageId:languageId result:result];
    }
}

// Identifies the possible languages for a given text.
// For each identified langauge a confidence value is returned as well.
// Read more here: https://developers.google.com/ml-kit/language/identification/ios
- (void)identifyPossibleLanguagesInText:(NSString *)text
                             languageId:(MLKLanguageIdentification*) languageId
                                 result:(FlutterResult)result {
    [languageId identifyPossibleLanguagesForText:text
                                      completion:^(NSArray * _Nonnull identifiedLanguages,
                                                   NSError * _Nullable error) {
        if (error != nil) {
            result(getFlutterError(error));
            return;
        }
        NSMutableArray *resultArray = [NSMutableArray array];
        for (MLKIdentifiedLanguage *language in identifiedLanguages) {
            NSDictionary *data = @{
                @"language" : language.languageTag,
                @"confidence" : [NSNumber numberWithFloat: language.confidence],
            };
            [resultArray addObject:data];
        }
        result(resultArray);
    }];
}

// Identify the language for a given text.
// Read more here: https://developers.google.com/ml-kit/language/identification/ios
- (void)identifyLanguageInText:(NSString *)text
                    languageId:(MLKLanguageIdentification*) languageId
                        result:(FlutterResult)result {
    [languageId identifyLanguageForText:text
                             completion:^(NSString * _Nonnull languageTag,
                                          NSError * _Nullable error) {
        if (error != nil) {
            result(getFlutterError(error));
            return;
        }
        result(languageTag);
    }];
}

@end

File: packages/google_mlkit_language_id/ios/Classes/GoogleMlKitLanguageIdPlugin.swift
import Flutter
import MLKitLanguageID
import google_mlkit_commons

@objc
public class GoogleMlKitLanguageIdPlugin: NSObject, FlutterPlugin {
  private var instances: [String: LanguageIdentification] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "google_mlkit_language_identifier",
      binaryMessenger: registrar.messenger()
    )
    let instance = GoogleMlKitLanguageIdPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "nlp#startLanguageIdentifier":
      handleDetection(call: call, result: result)
    case "nlp#closeLanguageIdentifier":
      if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
        instances.removeValue(forKey: uid)
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func initialize(call: FlutterMethodCall) -> LanguageIdentification? {
    guard let args = call.arguments as? [String: Any],
          let confidence = args["confidence"] as? NSNumber else { return nil }
    let options = LanguageIdentificationOptions(confidenceThreshold: confidence.floatValue)
    return LanguageIdentification.languageIdentification(options: options)
  }

  private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let uid = args["id"] as? String,
          let text = args["text"] as? String else {
      result(FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
      return
    }

    let languageId: LanguageIdentification
    if let existing = instances[uid] {
      languageId = existing
    } else {
      guard let newLanguageId = initialize(call: call) else {
        result(FlutterError(code: "invalid_args", message: "Invalid options", details: nil))
        return
      }
      languageId = newLanguageId
      instances[uid] = languageId
    }

    let possibleLanguages = (args["possibleLanguages"] as? NSNumber)?.boolValue ?? false
    if possibleLanguages {
      identifyPossibleLanguages(in: text, languageId: languageId, result: result)
    } else {
      identifyLanguage(in: text, languageId: languageId, result: result)
    }
  }

  private func identifyPossibleLanguages(
    in text: String,
    languageId: LanguageIdentification,
    result: @escaping FlutterResult
  ) {
    languageId.identifyPossibleLanguages(for: text) { identifiedLanguages, error in
      if let error = error as NSError? {
        result(FlutterError(
          code: "Error \(error.code)",
          message: error.domain,
          details: error.localizedDescription
        ))
        return
      }
      guard let list = identifiedLanguages else {
        result([])
        return
      }
      let resultArray = list.map { (lang: IdentifiedLanguage) -> [String: Any] in
        ["language": lang.languageTag, "confidence": NSNumber(value: lang.confidence)]
      }
      result(resultArray)
    }
  }

  private func identifyLanguage(
    in text: String,
    languageId: LanguageIdentification,
    result: @escaping FlutterResult
  ) {
    languageId.identifyLanguage(for: text) { languageTag, error in
      if let error = error as NSError? {
        result(FlutterError(
          code: "Error \(error.code)",
          message: error.domain,
          details: error.localizedDescription
        ))
        return
      }
      result(languageTag)
    }
  }
}
